### PR TITLE
Require keyword arguments for symmetry operations and atom basis information

### PR DESCRIPTION
**Pull request by g5t**
Along with other improvements with the move to `v0.6.0` this PR finishes implementing the suggestions in #42.

Most notably, by wrapping `brille._brille.Lattice` initialization in Python a user now must provide symmetry operation and/or atom basis information via keyword arguments. The provided keyword value can be either a fully-constructed `brille._brille.[X]` object *or* an appropriate tuple used to construct and initialize such an object. For `Lattice = brille.Lattice`, the supported keyword values are:
- `Lattice(..., symmetry=Symmetry_object, ...)` or `Lattice(..., symmetry=(rotations, translations), ...)`
- `Lattice(..., basis=Basis_object, ...)` or `Lattice(..., basis=(atom_positions, atom_type_indexes), ...)`

Information about the crystal system can be provided as before, as one or two position-based argument string(s), *or* as a keyword argument `spacegroup` which is either one string or a tuple of two strings. As before a single string should be one of:
- Hall symbol, encoding the generators of the symmetry operations of the crystal system
- Hermann-Mauguin symbol, as from the International Tables of Crystallography
- The generators or symmetry operations of the crystal system encoded in CIF xyz format
And two strings should be a Hermann-Mauguin symbol plus its origin shifting, axis orientation, etc., 'choice', as appropriate.

### CODE REVIEW

File: brille/utils.py [hunk 1]
```diff
@@ -109,9 +121,10 @@ def create_bz(*args, is_reciprocal=False, use_primitive=True, search_length=1,
 
     E.g. you cannot mix specifing `a`, `b`, `c`, and `angs` etc.
     """
+    from brille import Lattice
     # Take keyword arguments in preference to positional ones
     a, b, c, alpha, beta, gamma, lens, angs = (kwargs.pop(pname, None)
-        for pname in ['a', 'b', 'c', 'alpha', 'beta', 'gamma', 'lens', 'angs'])
+                                               for pname in ['a', 'b', 'c', 'alpha', 'beta', 'gamma', 'lens', 'angs'])
```
Comment (mducle) [inline brille/utils.py:127]: Since you've simplified the `Lattice` constructor to not allow single values, should we still support those in this helper `create_bz` function? (Although removing those now is more work and might break some scripts, it might make it easier to maintain in future?) 

On the other, should `create_bz` also now handle `basis` or `symmetry` as keyword arguments?

(I don't have any views either way ;)

**Review by mducle — APPROVED** [review]
Looks good... just had a few nitpicky comments...

(I haven't reviewed `subscript.hpp` because I have no idea what it's meant to do!)

Edit: I agree with the new simplified syntax where the single value `a, b, c, alpha, beta, gamma` syntax is removed.

File: brille/lattice.py [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+        if len(args) > 2 and isinstance(args[-1], str) and isinstance(args[-2], str):
+            spacegroup = args[-2], args[-1]
+            args = args[:-2]
+        elif len(args) > 1 and isinstance(args[-1], str):
+            spacegroup = args[-1],
+            args = args[:-1]
+        else:
+            spacegroup = tuple()
+
+    if symmetry is None and basis is None:
```
Comment (rebeccafair) [inline brille/lattice.py:85]: The if logic here is slightly hard to follow maybe it could be rearranged to be clearer?

e.g. maybe get the errors out of the way first?
```
if symmetry provided and spacegroup provided
# error
if basis and not symmetry provided and not spacegroup provided
# error
```
Also here we're allowing a spacegroup to be provided without a basis, is that ok?

Are all these code paths tested?

File: brille/lattice.py [hunk 3]
```diff
@@ -0,0 +1,97 @@
+try:
+    from ._brille import Lattice as _Lattice, Symmetry as _Symmetry, Basis as _Basis
+except ModuleNotFoundError:
+    # This allows CTest-based testing which is done before creating/installing the wheel
+    from pathlib import Path
+    from .load import load
+    _brille = load(['_brille'], search=[Path(), Path('..')])
+    _Lattice, _Symmetry, _Basis = [getattr(_brille, x) for x in ('Lattice', 'Symmetry', 'Basis')]
```
Comment (rebeccafair) [inline brille/lattice.py:8]: I'm not sure I like testing workarounds being in the distributed package, but I guess unless wheels are built and installed before testing, there's no other option?

File: brille/utils.py (hunk 1 above)
Comment (rebeccafair) [reply brille/utils.py:127]: I think `create_bz` should maybe even just take `Lattice` as a required positional argument, to avoid a lot of the cell/spacegroup/symmetry logic that is repeated here. But would that make this helper function obsolete?

I think we should lean towards giving users fewer ways of doing the same things as it's easier to maintain but that's just me...

**Review by rebeccafair — CHANGES_REQUESTED** [review]
This will definitely make things nicer to use from Python! Just a few comments

File: brille/lattice.py [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+        If present, the 'spacegroup' positional argument(s) or keyword must not be used.
+    basis: brille.Basis, tuple(positions, types)
+        The atom basis information of the lattice, expressed in units of the real space basis vectors. The types
+        must be integer and are used only to identify equivalent atoms -- they should probably be contiguous from
+        zero to 1-N where N is the number of unique atoms in the atom basis.
+        If present, either spacegroup or symmetry information must be provided.
+    """
+    if spacegroup is None:
+        # look for a string (or two!) on the end of the position arguments
+        if len(args) > 2 and isinstance(args[-1], str) and isinstance(args[-2], str):
```
[the remark this answers is not in the document]
Comment (g5t) [reply brille/lattice.py:76]: @rebeccafair also suggested removing non-keyword spacegroup information, so I'll capitulate.

I'll add the same into the docstring, but for completeness:
```
Valid syntax for (Hermann-Mauguin, choice) input depends on the spacegroup but is generally one of:
    a single letter ('a', 'b', or 'c') with possible prepended '-', denoting unique-axis choice
    a single digit ('1', '2', or '3'), denoting origin choice
    a letter and digit, denoting unique-axis and origin choice
    a permutation of 'abc' with possible '-' before one of the letters, denoting axis permutation
    or 'R' or 'H' for trigonal systems with Rhombohedral or Hexagonal lattice settings, respectively.
Acceptable values are contained in the C++ source code in the seventh column of the table
https://github.com/brille/brille/blob/eecb4cb28227665908793abc47e88c69518c09fc/src/spg_database.cpp#L63-L610
with each line representing one spacegroup with values, in order, defined by the class signature at
https://github.com/brille/brille/blob/eecb4cb28227665908793abc47e88c69518c09fc/src/spg_database.hpp#L80-L91
These values come from spglib, which likely obtained them from Seto's now defunct Home Page,
https://web.archive.org/web/20210621195003/http://pmsl.planet.sci.kobe-u.ac.jp/~seto/?page_id=37&lang=en
```

File: brille/utils.py (hunk 1 above)
Comment (g5t) [reply brille/utils.py:127]: I would be happy to remove the scalar input parsing from `create_bz`, I've left it at present to avoid breaking possible use cases.
Changing `create_bz` to require a pre-constructed `brille._brille.Lattice` is fine for me, but may be a step too far for others? Ignoring the scalar value constructors, the syntax would change from
```python
# positional argument(s) for lattice parameters
bz = create_bz(lengths, angles, **lattice_keywords, **bz_keywords)
bz = create_bz(vectors, **lattice_keywords, **bz_keywords)
# equivalent keyword arguments
bz = create_bz(lens=lengths, angs=angles, **lattice_keywords, **bz_keywords)
bz = create_bz(lattice_vectors=vectors, **lattice_keywords, **bz_keywords)
```
to
```python
bz = create_bz(Lattice(lengths, angles, **lattice_keywords), **bz_keywords)
bz = create_bz(Lattice(vectors, **lattice_keywords), **bz_keywords)
```
where the keyword-argument variants will no longer work since the signature of the `Lattice` wrapper is
`def Lattice(*args, **kwargs)`. Keeping both the position and keyword argument syntaxes requires functionality like `create_bz`, though I'm not sure if both versions are used.

Does anyone use `create_bz(**only_keywords)`?

File: brille/lattice.py [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+    symmetry: brille.Symmetry, tuple(matrices, vectors), str
+        The spacegroup symmetry operations as an object, a tuple of the (pseudo)rotation matrices and translation
+        vectors, or a CIF xyz encoded string. The symmetry information must be provided by keyword.
+        If present, the 'spacegroup' positional argument(s) or keyword must not be used.
+    basis: brille.Basis, tuple(positions, types)
+        The atom basis information of the lattice, expressed in units of the real space basis vectors. The types
+        must be integer and are used only to identify equivalent atoms -- they should probably be contiguous from
+        zero to 1-N where N is the number of unique atoms in the atom basis.
+        If present, either spacegroup or symmetry information must be provided.
+    """
```
[the remark this answers is not in the document]
Comment (g5t) [reply brille/lattice.py:73]: I can add the current crop of keyword arguments to the docstring, but I was hoping to avoid copying signatures from the pybind11 wrappers -- especially in case those signatures change in the future.
I'll look into providing links in the online documentation.

File: brille/lattice.py [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+    ----
+    The parameters packed into `args` are position-based and should be *either* `(lengths, angles)`
+    *or* `(vectors)` [which also requires setting `row_vectors=False` if the matrix represents column vectors].
+
+
+    Parameters
+    ----------
+    args:
+        lengths : list, tuple, numpy.ndarray
+            The three basis vector lengths in angstrom for real lattice, or inverse angstrom for reciprocal lattices
```
[the remark this answers is not in the document]
Comment (g5t) [reply brille/lattice.py:51]: Via one of the keyword arguments not documented here, of course :sweat:

File: brille/lattice.py [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+        vectors, or a CIF xyz encoded string. The symmetry information must be provided by keyword.
+        If present, the 'spacegroup' positional argument(s) or keyword must not be used.
+    basis: brille.Basis, tuple(positions, types)
+        The atom basis information of the lattice, expressed in units of the real space basis vectors. The types
+        must be integer and are used only to identify equivalent atoms -- they should probably be contiguous from
+        zero to 1-N where N is the number of unique atoms in the atom basis.
+        If present, either spacegroup or symmetry information must be provided.
+    """
+    if spacegroup is None:
+        # look for a string (or two!) on the end of the position arguments
```
[the remark this answers is not in the document]
Comment (g5t) [reply brille/lattice.py:75]: Good suggestion, this is changed to keyword-only and will be pushed shortly.

File: brille/lattice.py [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+
+    Note
+    ----
+    The parameters packed into `args` are position-based and should be *either* `(lengths, angles)`
+    *or* `(vectors)` [which also requires setting `row_vectors=False` if the matrix represents column vectors].
+
+
+    Parameters
+    ----------
+    args:
```
[the remark this answers is not in the document]
Comment (g5t) [reply brille/lattice.py:49]: Avoiding nonsensical cases is a good idea. I'm willing to give the single positional argument idea a chance. I've tentatively gone with `values` instead of `cell`, since I think the latter has a specific meaning for `spglib` and `euphonic` users.

Should I make the single argument positional-only? E.g., `def Lattice(values, /, **kwargs)`.
Then a user can't provide the argument as a keyword, so its name is less important?

File: brille/lattice.py (hunk 2 above)
Comment (g5t) [reply brille/lattice.py:85]: :+1:  Thanks for the suggestion, I'll implement it.

User provided spacegroup string(s) or symmetry operations without an atom basis are fine but a basis without either is not.

File: brille/lattice.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (rebeccafair) [reply brille/lattice.py:49]: Yeah, I think if we _always_ need an argument, it should be positional, good point about its name being less important in that case.

File: brille/lattice.py (hunk 3 above)
Comment (rebeccafair) [reply brille/lattice.py:8]: I've never used it, but could [\_\_debug\_\_](https://docs.python.org/3/library/constants.html#debug__) be helpful here? It looks like you can run Python with the [-O](https://docs.python.org/3/using/cmdline.html#cmdoption-o) option to run any code in a `__debug__` section. Or have you tried an editable install with `pip install -e`? Sorry if these suggestions are no good, I'm not so familiar with how the tests are run.

I don't imagine the try/catch will have an effect on performance, but I worry that if a user doesn't have `brille` installed properly, it could trigger the `catch` and hide what the real problem is.

Comment (g5t) [reply brille/lattice.py:8]: I'd guess `pip install -e` is similar to the `develop` option for `setup.py` -- this is a really good way to avoid re-building the compiled module for every change to the Python code, but the package is still installed (via an 'egg link') for the current Python interpreter.
If that interpreter is inside of a development-specific virtual environment ([Virualenv](https://virtualenv.pypa.io/en/latest/), Conda-like, [pipenv](https://pipenv.pypa.io/en/latest/), [poetry](https://python-poetry.org/), or similar) there's no problem. But if a developer (me) accidentally installs the development module for, e.g., the system interpreter, there is the possiblity to unintentionally use the development instead of stable version of the module elsewhere.

I'll try to think of a simpler way (e.g., something I can't forget to do when building `brille`) to avoid this while also not putting an unnecessary `try/except` block in the distributed module.

File: brille/lattice.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (g5t) [reply brille/lattice.py]: Position-only arguments are not available in Python 3.6 and 3.7 so I _can not_ make this so strict until the minimum supported version of Python can move to 3.8.

Python 3.6 is past EOL and 3.7 will pass the same milestone in 11 months. Does PACE still plan to support 3.6 and 3.7?

[the remark this answers is not in the document]
Comment (rebeccafair) [reply brille/lattice.py]: Sorry, I didn't really mean positional argument, I meant required! So the Python versioning shouldn't be an issue (sorry, it's bad of me to muddy those 2 things).

But FYI Euphonic will stop supporting Python 3.6 as of Euphonic 1.0, and from now on we'll probably attempt to roughly follow whatever Python versions have official support (unless there's a reason not to).

File: brille/lattice.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (g5t) [reply brille/lattice.py:86]: This is done in the docstring but other documentation changes are needed, so #84 will pull the changes into the built HTML file.

File: brille/lattice.py (hunk 3 above)
Comment (g5t) [reply brille/lattice.py]: The `try/except` blocks are now gone and developer tools have been added to build and test the CTest and python tests in a PyPA manylinux image.
